keps/sig-api-machinery/20191210-consistent-reads-from-cache.md

```diff
@@ -0,0 +1,455 @@
+
+- Get the current revision from etcd for the resource type being served. The returned revision is strongly consistent (guaranteed to be the latest revision via a quorum read).
+- Use the existing `waitUntilFreshAndBlock` function in the watch cache to wait briefly for the watch to catch up to the current revision.
+- If the block times out, fail the request (see "What if the watch cache is stale?" section for details)
+
+To get the revsion we have some options: 
+
+- Use an etcd range request with `WithCount` enabled so etcd return only a count and revision 
+- Use an etcd range request against a known empty range with limit=1 as an additional guard (since etcd does not allow for limit=0)
+
+Consistent GET requests will continue to be served directly from etcd. We will
+only serve consistent LIST requests from cache.
+
+Optional: For some (but not all) of the etcd progress watch events, also create a
+kubernetes "bookmark" watch event and send it to kube-apiserver clients so that
+reflectors and shared informers are kept up-to-date. The benefit of this is that
+it minimizes the chance that these clients will end up with an out-of-date
+resource version and need to relist (which can impact scalability). See [Watch
+Bookmarks](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190206-watch-bookmark.md)
+for details.
+
+#### Determining if etcd is sending progress notify events
+
+It is possible to automatically determine if etcd is sending progress notify events.
+Each watch cache could keep track of when it received the last progress notify event.
+If it has been sufficiently long since the last one was received, or if none have
+ever been received, the watch cache should assume etcd is not sending progress notify,
+and not attempt to serve consistent reads from cache, falling back to serving
+the reads directly from etcd.
+
+### Risks and Mitigations
+
+Configuring etcd to send progress notify events may have performance implications.
+@mm4tt and @wojtek-t have run [some
+experiments](https://github.com/kubernetes/kubernetes/pull/86769) that suggest
+that a progress notify interval of 50ms results in a noticeable increase in both
+etcd and kube-apiserver CPU utilization and some increased kube-apiserver serving
+latency. We intend to address this by first trying a 250ms progress notify interval,
+if ths performs well we will use that as our interval. If not, we will dig in
+to figure out why the events are impacting performance and see if we can optimize
+it away.
+
+## Design Details
+
+### Progress notify interval selection
+
+Not all LIST requests require consistency, and we're taking the view that if you
+really want to have a consistent LIST (we should explicitly exclude GETs from
+it), then you may need to pay additional tax latency for it. For this reason, we
+intend to start with a 250ms progress notify interval, which will on average 125ms
```

Comment:
This tax seems pretty significant.  Please also mention it in a more prominent section.

Reply:
I'll highlight it. Also, @jfbai's proposal for using the etcd `WithLastRev` request option might eliminate this tax. We'll open a follow up PR to explore that once this is merged provisional.